File: converts.js
const express = require('express');
const { initializeApp, cert } = require('firebase-admin/app');
const { getFirestore, collection, where, getDocs, addDoc, doc, updateDoc, arrayUnion } = require('firebase-admin/firestore');
const bcrypt = require('bcrypt');
const cors = require('cors');

const serviceAccount = require('./capstone.json');

// Initialize Firebase App first
initializeApp({
  credential: cert(serviceAccount)
});
// Initialize Firestore after Firebase App is initialized
const db = getFirestore();

// Continue with the rest of your code


const app = express();
app.use(cors());
app.use(express.static('public'));
app.use(express.urlencoded({ extended: true }));
app.use(express.json());

app.get('/login', (req, res) => {
  res.sendFile(__dirname + '/public/' + 'logincrop.html');
});

app.post('/signup', async (req, res) => {
  const { email, password } = req.body;
    try {
      // Query Firestore to check if a user with the same email already exists
      const userRef = db.collection('users'); // Corrected this line
      const querySnapshot = await userRef.where('email', '==', email).get();
  
      if (!querySnapshot.empty) {
        res.status(409).send('User with this email already exists');
      } else {
        // Store user data in Firestore
        const hashedPassword = await bcrypt.hash(password, 10);
  
        await userRef.add({
          email: email,
          password: hashedPassword,
        });
  
        // Redirect to the indexed.html page after successful signup
        res.redirect('/indexed.html');
      }
    } catch (error) {
      res.status(500).send('Error signing up: ' + error.message);
    }
  });
  
  app.post('/loginsubmit', async (req, res) => {
    const { email, password } = req.body;
  
    try {
      // Query Firestore to check user credentials
      const userRef = db.collection('users'); // Corrected this line
      const querySnapshot = await userRef.where('email', '==', email).get();
  
      if (querySnapshot.empty) {
        res.status(401).send('Invalid credentials');
      } else {
        const userData = querySnapshot.docs[0].data();
        const hashedPassword = userData.password;
  
        // Compare the hashed password
        const passwordMatch = await bcrypt.compare(password, hashedPassword);
        if (passwordMatch) {
          // Successful login
          res.redirect('/indexed.html');
        } else {
          res.status(401).send('Invalid credentials');
        }
      }
    } catch (error) {
      res.status(500).send('Error checking credentials: ' + error.message);
    }
  });
  app.post('/predict', async (req, res) => {
    // Extract the input data from the request body
    const { nitrogen, phosphorus, potassium, temperature, humidity, ph, rainfall } = req.body;
  
    // Prepare the data to send to the Flask server
    const data = {
      nitrogen,
      phosphorus,
      potassium,
      temperature,
      humidity,
      ph,
      rainfall
    };
  
    try {
      const flaskResponse = await fetch('http://localhost:5000/predict', {
        method: 'POST',
        body: JSON.stringify(data),
        headers: { 'Content-Type': 'application/json' }
      });
      const prediction = await flaskResponse.text();

    res.send(prediction);
  } catch (error) {
    res.status(500).send('Error getting prediction: ' + error.message);
  }
});
  app.listen(3000, () => {
    console.log('App is listening on port 3000');
  });

File: converts.py
from flask import Flask, render_template, request
import pandas as pd
import numpy as np
import pickle
import warnings

app = Flask(__name__, template_folder='public')
warnings.filterwarnings("ignore", message="Trying to unpickle estimator")

def load_model(modelfile):
    loaded_model = pickle.load(open(modelfile, 'rb'))
    return loaded_model

@app.route('/')
def index():
    return render_template('indexed.html')

@app.route('/predict', methods=['POST'])
def predict():
    if request.method == 'POST':
        N = float(request.form['nitrogen'])
        P = float(request.form['phosphorus'])
        K = float(request.form['potassium'])
        temp = float(request.form['temperature'])
        humidity = float(request.form['humidity'])
        ph = float(request.form['ph'])
        rainfall = float(request.form['rainfall'])

        feature_list = [N, P, K, temp, humidity, ph, rainfall]
        single_pred = np.array(feature_list).reshape(1, -1)

        loaded_model = load_model('C:/Users/anush/bharath intern/group/model (3).pkl')  # Update the path to your model file
        prediction = loaded_model.predict(single_pred)

        # Convert the prediction to a string before using title()
        prediction_str = str(prediction)

        return render_template('indexed.html', prediction=prediction_str.title())

if __name__ == '__main__':
    app.run(debug=True)
